src/window.c

```diff
@@ -11,19 +11,33 @@ See the file LICENSE for details.
 #define CHARACTER_W 8
 #define CHARACTER_H 8
 
-static int x_offset = 0;
-static int y_offset = 0;
+static inline void window_clear_bounds(struct window *win) {
+    win->bounds_x_1 = 0;
+    win->bounds_x_2 = 0;
+    win->bounds_y_1 = 0;
+    win->bounds_y_2 = 0;
+}
+
+static inline void window_set_bounds(struct window *win, int x, int y, int w, int h) {
+    // check if bounds_x_1 ... y_2 are -1.
+    // if so, do calculation. otherwise, do nothing
+    win->bounds_x_1 = x < 0 ? 0 : x;
+    win->bounds_x_2 = x + w >= graphics_width() ? graphics_width() - 1 : x + w;
+    win->bounds_y_1 = y < 0 ? 0 : y;
+    win->bounds_y_2 = y + h >= graphics_height() ? graphics_height() - 1 : y + h;
+}
 
 // This function prepares to draw in a window by setting up its boundary
 // restrictions.
-// THIS FUNCTION SHOULD ALWAYS BE CALLED BEFORE MAKING A GRAPHICS DRAW
-// CALL IN THE WINDOW SYSTEM
-static inline void window_begin_draw(struct window *w) {
+// THIS FUNCTION SHOULD ALWAYS BE CALLED WHEN MAKING A WINDOW
+// OR MOVING A WINDOW
+static inline void window_init_draw(struct window *w) {
```

Comment:
begin to me implies that there should be an end. init does not say the same thing. It just says that we initialize the window's ability to draw.